异常是不可控的，会影响最终的呈现结果，但是我们有充分的理由去做这样的事情。

## 常见的异常
* JS 语法错误、代码异常
* AJAX 请求异常
* 静态资源加载异常
* Promise 异常
* Iframe 异常
* 跨域 Script error
* 崩溃和卡顿

## Try-Catch 的误区
try-catch 只能捕获到同步的运行时错误，对语法和异步错误却无能为力，捕获不到。
### 同步运行时错误
```
try {
  let name = 'jartto';
  console.log(nam);
} catch(e) {
  console.log('捕获到异常：',e);
}
捕获到异常： ReferenceError: nam is not defined
    at <anonymous>:3:15
```
### 不能捕获到语法错误
```
try {
  let name = 'jartto; // 删除一个单引号
  console.log(nam);
} catch(e) {

  console.log('捕获到异常：',e);
}

Uncaught SyntaxError: Invalid or unexpected token
```
### 异步错误
```
try {
  setTimeout(() => {
    undefined.map(v => v);
  }, 1000)
} catch(e) {
  console.log('捕获到异常：',e);
}

Uncaught TypeError: Cannot read property 'map' of undefined
    at setTimeout (<anonymous>:3:11)
```

## 引用
[如何优雅处理前端异常？](http://jartto.wang/2018/11/20/js-exception-handling/)
